x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvSlice.java

```diff
@@ -0,0 +1,315 @@
+                toEvaluator.apply(end)
+            );
+            case LONG -> new MvSliceLongEvaluator.Factory(
+                source(),
+                toEvaluator.apply(field),
+                toEvaluator.apply(start),
+                toEvaluator.apply(end)
+            );
+            case NULL -> EvalOperator.CONSTANT_NULL_FACTORY;
+            default -> throw EsqlIllegalArgumentException.illegalDataType(field.dataType());
+        };
+    }
+
+    @Override
+    public Object fold() {
+        return EvaluatorMapper.super.fold();
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new MvSlice(source(), newChildren.get(0), newChildren.get(1), newChildren.size() > 2 ? newChildren.get(2) : null);
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, MvSlice::new, field, start, end);
+    }
+
+    @Override
+    public DataType dataType() {
+        return field.dataType();
+    }
+
+    @Override
+    public ScriptTemplate asScript() {
+        throw new UnsupportedOperationException("functions do not support scripting");
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(field, start, end);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null || obj.getClass() != getClass()) {
+            return false;
+        }
+        MvSlice other = (MvSlice) obj;
+        return Objects.equals(other.field, field) && Objects.equals(other.start, start) && Objects.equals(other.end, end);
+    }
+
+    static int calculateOffset(int oldOffset, int fieldValueCount) {
+        return oldOffset < 0 ? oldOffset + fieldValueCount : oldOffset;
+    }
+
+    static void checkStartEnd(int start, int end) throws InvalidArgumentException {
+        if (start > end) {
+            throw new InvalidArgumentException("Start offset is greater than end offset");
+        }
+        if (start < 0 && end >= 0) {
+            throw new InvalidArgumentException("Start and end offset have different signs");
+        }
+    }
+
+    @Evaluator(extraName = "Boolean", warnExceptions = { InvalidArgumentException.class })
+    static void process(BooleanBlock.Builder builder, int position, BooleanBlock field, int start, int end) {
+        int fieldValueCount = field.getValueCount(position);
+        checkStartEnd(start, end);
+        int first = field.getFirstValueIndex(position);
+        int startOffset = calculateOffset(start, fieldValueCount);
+        int endOffset = calculateOffset(end, fieldValueCount);
+        if (startOffset > fieldValueCount) {
+            builder.appendNull();
+            return;
+        }
+        builder.beginPositionEntry();
+        for (int i = startOffset + first; i < Math.min(fieldValueCount, endOffset + 1) + first; i++) {
+            builder.appendBoolean(field.getBoolean(i));
+        }
+        builder.endPositionEntry();
```

Comment:
`builder.beginPositionEntry()` and `builder.endPositionEntry()` add quite some overhead to the process, I think we should avoid invoking them when `start == end`